core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/progress/listener/ProgressUpdater.java

```diff
@@ -0,0 +1,243 @@
+import software.amazon.awssdk.core.internal.progress.ProgressListenerFailedContext;
+import software.amazon.awssdk.core.internal.progress.snapshot.DefaultProgressSnapshot;
+import software.amazon.awssdk.core.progress.listener.SdkRequestProgress;
+import software.amazon.awssdk.core.progress.snapshot.ProgressSnapshot;
+
+/**
+ * ProgressUpdater exposes methods that invokes listener methods to update and store request progress state
+ */
+@SdkInternalApi
+public class ProgressUpdater {
+    private final DefaultSdkRequestProgress uploadProgress;
+    private final DefaultSdkRequestProgress downloadProgress;
+    private final ProgressListenerContext context;
+    private final ProgressListenerInvoker listenerInvoker;
+    private final CompletableFuture<Void> endOfStreamFuture;
+
+    public ProgressUpdater(SdkRequest sdkRequest,
+                           Long contentLength) {
+        DefaultProgressSnapshot.Builder uploadProgressSnapshotBuilder = DefaultProgressSnapshot.builder();
+        uploadProgressSnapshotBuilder.transferredBytes(0L);
+        Optional.ofNullable(contentLength).ifPresent(uploadProgressSnapshotBuilder::totalBytes);
+
+        ProgressSnapshot uploadProgressSnapshot = uploadProgressSnapshotBuilder.build();
+        uploadProgress = new DefaultSdkRequestProgress(uploadProgressSnapshot);
+
+        DefaultProgressSnapshot.Builder downloadProgressSnapshotBuilder = DefaultProgressSnapshot.builder();
+        downloadProgressSnapshotBuilder.transferredBytes(0L);
+        Optional.ofNullable(contentLength).ifPresent(downloadProgressSnapshotBuilder::totalBytes);
+
+        ProgressSnapshot downloadProgressSnapshot = downloadProgressSnapshotBuilder.build();
+        downloadProgress = new DefaultSdkRequestProgress(downloadProgressSnapshot);
+
+        context = ProgressListenerContext.builder()
+                                         .request(sdkRequest)
+                                         .uploadProgressSnapshot(uploadProgressSnapshot)
+                                         .downloadProgressSnapshot(downloadProgressSnapshot)
+                                         .build();
+
+        listenerInvoker = sdkRequest.overrideConfiguration().get().progressListeners() == null
+                          ? new ProgressListenerInvoker((Collections.emptyList()))
+                          : new ProgressListenerInvoker(sdkRequest.overrideConfiguration().get().progressListeners());
+
+        endOfStreamFuture = new CompletableFuture<>();
+    }
+
+    public SdkRequestProgress uploadProgress() {
+        return uploadProgress;
+    }
+
+    public SdkRequestProgress downloadProgress() {
+        return downloadProgress;
+    }
+
+    public AsyncRequestBody wrapUploadRequestBody(AsyncRequestBody requestBody) {
+        return AsyncRequestBodyListener.wrap(
+            requestBody,
+            new AsyncRequestBodyListener() {
+                final AtomicBoolean done = new AtomicBoolean(false);
```

Comment:
I understand we have to use a flag because onComplete may not be invoked. Can we add a note here to mention this https://github.com/aws/aws-sdk-java-v2/pull/4440